File: objectModel/TypeScript/__test__/Persistence/CdmFolder/JsonSemanticVersion.test.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import {
    CdmCorpusDefinition,
    cdmStatusLevel
} from '../../../internal';

import { testHelper } from '../../testHelper';

describe('Persistence.CdmFolder.JsonSemanticVersionTest', () => {
    /**
     * The path between TestDataPath and TestName.
     */
    const testsSubpath: string = 'Persistence/CdmFolder/JsonSemanticVersionTest';

    /**
     * Test loading a document with a semantic version bigger than the one supported.
     */
    it('testLoadingUnsupportedVersion', async (done) => {
        const corpus: CdmCorpusDefinition = testHelper.getLocalCorpus(testsSubpath, 'TestLoadingUnsupportedVersion');
        let errorCount: number = 0;

        // Test loading a resolved document.
        corpus.setEventCallback((level: cdmStatusLevel, message: string) => {
            if (message.includes('This ObjectModel version supports json semantic version') && level == cdmStatusLevel.warning) {
                errorCount++;
            }
        }, cdmStatusLevel.warning);
        await corpus.fetchObjectAsync('local:/resolvedDoc.cdm.json');
        if (errorCount !== 1) {
            done.fail('Should have logged a warning.');
        }
        errorCount = 0;

        // Test loading a logical document.
        corpus.setEventCallback((level: cdmStatusLevel, message: string) => {
            if (message.includes('This ObjectModel version supports json semantic version') && level == cdmStatusLevel.error) {
                errorCount++;
            }
        }, cdmStatusLevel.error);
        await corpus.fetchObjectAsync('local:/logicalDoc.cdm.json');
        if (errorCount !== 1) {
            done.fail('Should have logged an error.');
        }
        errorCount = 0;

        // Test loading a document missing the jsonSemanticVersion property.
        corpus.setEventCallback((level: cdmStatusLevel, message: string) => {
            if (message.includes('jsonSemanticVersion is a required property of a document.') && level == cdmStatusLevel.warning) {
                errorCount++;
            }
        }, cdmStatusLevel.warning);
        await corpus.fetchObjectAsync('local:/missingDoc.cdm.json');
        if (errorCount !== 1) {
            done.fail('Should have logged a warning for missing property.');
        }
        done();
    });

    /**
     * Test loading a document with a semantic version bigger than the one supported.
     */
    it('testLoadingInvalidVersion', async (done) => {
        const corpus: CdmCorpusDefinition = testHelper.getLocalCorpus(testsSubpath, 'TestLoadingInvalidVersion');
        let errorCount: number = 0;

        corpus.setEventCallback((level: cdmStatusLevel, message: string) => {
            if (message.includes('jsonSemanticVersion must be set using the format <major>.<minor>.<patch>.') && level == cdmStatusLevel.warning) {
                errorCount++;
            }
        }, cdmStatusLevel.warning);

        // Test loading a version format "a.0.0".
        await corpus.fetchObjectAsync('local:/invalidVersionDoc.cdm.json');
        if (errorCount !== 1) {
            done.fail('Should have logged a warning.');
        }
        errorCount = 0;

        // Test loading a version format "1.0".
        await corpus.fetchObjectAsync('local:/invalidFormatDoc.cdm.json');
        if (errorCount !== 1) {
            done.fail('Should have logged a warning.');
        }
        done();
    });
});

File: objectModel/Java/objectmodel/src/test/java/com/microsoft/commondatamodel/objectmodel/persistence/cdmfolder/JsonSemanticVersionTest.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder;

import java.io.File;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.atomic.AtomicInteger;

import com.microsoft.commondatamodel.objectmodel.TestHelper;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmCorpusDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmDocumentDefinition;
import com.microsoft.commondatamodel.objectmodel.enums.CdmStatusLevel;

import org.testng.Assert;
import org.testng.annotations.Test;

public class JsonSemanticVersionTest {
  /**
   * The path between TestDataPath and TestName.
   */
  private static final String TESTS_SUBPATH = new File(new File("persistence", "cdmfolder"), "JsonSemanticVersionTest")
      .toString();

  /**
   * Test loading a document with a semantic version bigger than the one supported.
   */
  @Test
  public void testLoadingUnsupportedVersion() throws InterruptedException, ExecutionException {
    CdmCorpusDefinition corpus = TestHelper.getLocalCorpus(TESTS_SUBPATH, "TestLoadingUnsupportedVersion");
    final AtomicInteger errorCount = new AtomicInteger(0);

    // Test loading a resolved document.
    corpus.setEventCallback((CdmStatusLevel level, String message) -> {
      if (message.contains("This ObjectModel version supports json semantic version") && level == CdmStatusLevel.Warning) {
        errorCount.getAndIncrement();
      }
    }, CdmStatusLevel.Warning);
    corpus.<CdmDocumentDefinition>fetchObjectAsync("local:/resolvedDoc.cdm.json").get();
    if (errorCount.get() != 1) {
      Assert.fail("Should have logged a warning.");
    }
    errorCount.set(0);

    // Test loading a logical document.
    corpus.setEventCallback((CdmStatusLevel level, String message) -> {
      if (message.contains("This ObjectModel version supports json semantic version") && level == CdmStatusLevel.Error) {
        errorCount.getAndIncrement();
      }
    }, CdmStatusLevel.Warning);
    corpus.<CdmDocumentDefinition>fetchObjectAsync("local:/logicalDoc.cdm.json").get();
    if (errorCount.get() != 1) {
      Assert.fail("Should have logged an error.");
    }
    errorCount.set(0);

    // Test loading a document missing the jsonSemanticVersion property.
    corpus.setEventCallback((CdmStatusLevel level, String message) -> {
      if (message.contains("jsonSemanticVersion is a required property of a document.") && level == CdmStatusLevel.Warning) {
        errorCount.getAndIncrement();
      }
    }, CdmStatusLevel.Warning);
    corpus.<CdmDocumentDefinition>fetchObjectAsync("local:/missingDoc.cdm.json").get();
    if (errorCount.get() != 1) {
      Assert.fail("Should have logged a warning for missing property.");
    }
  }

  /**
   * Test loading a document with an invalid semantic version.
   */
  @Test
  public void testLoadingInvalidVersion() throws InterruptedException, ExecutionException {
    CdmCorpusDefinition corpus = TestHelper.getLocalCorpus(TESTS_SUBPATH, "TestLoadingInvalidVersion");
    final AtomicInteger errorCount = new AtomicInteger(0);

    corpus.setEventCallback((CdmStatusLevel level, String message) -> {
      if (message.contains("jsonSemanticVersion must be set using the format <major>.<minor>.<patch>.") && level == CdmStatusLevel.Warning) {
        errorCount.getAndIncrement();
      }
    }, CdmStatusLevel.Warning);

    // Test loading a version format "a.0.0".
    corpus.<CdmDocumentDefinition>fetchObjectAsync("local:/invalidVersionDoc.cdm.json").get();
    if (errorCount.get() != 1) {
      Assert.fail("Should have logged a warning.");
    }
    errorCount.set(0);

    // Test loading a version format "1.0".
    corpus.<CdmDocumentDefinition>fetchObjectAsync("local:/invalidFormatDoc.cdm.json").get();
    if (errorCount.get() != 1) {
      Assert.fail("Should have logged a warning.");
    }
    errorCount.set(0);
  }
}